OpenRA.Mods.Common/UpdateRules/Rules/20180923/ExtractHackyAIModules.cs

```diff
@@ -55,18 +62,33 @@ public override IEnumerable<string> UpdateActorNode(ModData modData, MiniYamlNod
 
 			var addNodes = new List<MiniYamlNode>();
 
-			// We add a 'default' HarvesterBotModule in any case,
+			// We add a 'default' HarvesterBotModule in any case (unless the file doesn't contain any HackyAI base definition),
 			// and only add more for AIs that define custom values for one of its fields.
 			var defaultHarvNode = new MiniYamlNode("HarvesterBotModule", "");
+			var addDefaultHarvModule = false;
 
 			foreach (var hackyAINode in hackyAIs)
 			{
-				// HackyAIInfo.Name might contain spaces, so Type is better suited to be used as condition name
-				var aiType = hackyAINode.LastChildMatching("Type").NodeValue<string>();
+				// HackyAIInfo.Name might contain spaces, so Type is better suited to be used as condition name.
+				// Type can be 'null' if the place we're updating is overriding the default rules (like a map's rules.yaml).
+				// If that's the case, it's better to not perform most of the updates on this particular yaml file,
+				// as most - or more likely all - necessary updates will already have been performed on the base ai yaml.
+				var aiTypeNode = hackyAINode.LastChildMatching("Type");
```

Comment:
Most of the problems here come from trait *removals*.  It might make sense to pass `includeRemovals: false` to the `hackyAIs` query above?